web/source/dom/TouchAliasElement.ts

```diff
@@ -0,0 +1,648 @@
+// References the KMW string extensions.
+/// <reference path="../kmwstring.ts" />
+// References extra HTML definitions not included by default in TS.
+/// <reference path="../kmwexthtml.ts" />
+// References device-specific code checks (separable module from KMW)
+/// <reference path="../kmwdevice.ts" />
+// References common DOM utility functions (separate module from KMW)
+/// <reference path="utils.ts" />
+
+namespace com.keyman.dom {
+  /**
+   * As our touch-alias elements have historically been based on <div>s, this
+   * defines the root element of touch-aliases as a merger type with HTMLDivElements.
+   * 
+   */
+  export type TouchAliasElement = HTMLDivElement & TouchAliasData;
+
+  // Many thanks to https://www.typescriptlang.org/docs/handbook/advanced-types.html for this.
+  function link(elem: HTMLDivElement, data: TouchAliasData): TouchAliasElement {
+    let e = <TouchAliasElement> elem;
+    
+    // Merges all properties and methods of KeyData onto the underlying HTMLDivElement, creating a merged class.
+    for(let id in data) {
+      if(!e.hasOwnProperty(id)) {
+        (<any>e)[id] = (<any>data)[id];
+      }
+    }
+
+    return e;
+  }
+
+  export function constructTouchAlias(base?: HTMLElement): TouchAliasElement {
+    let div = document.createElement("div");
+    let ele = link(div, new TouchAliasData());
+
+    if(base) {
+      ele.initWithBase(base);
+    } else {
+      ele.init();
+    }
+
+    return ele;
+  }
+
+  /**
+   * The core definition for touch-alias 'subclassing' of HTMLDivElement.
+   * It's 'merged' with HTMLDivElement to avoid issues with DOM inheritance and DOM element creation.
+   */
+  class TouchAliasData {
+    ['base']: HTMLElement = null; // NOT undefined; we can use this distinction for 'type-checking'.
+    __preCaret:  HTMLSpanElement;
+    __postCaret: HTMLSpanElement;
+    __scrollDiv: HTMLDivElement;
+    __scrollBar: HTMLDivElement;
+
+    __caretSpan: HTMLSpanElement;
+    __caretDiv: HTMLDivElement;
+    __caretTimerId: number;
+    __activeCaret: boolean = false;
+
+    __resizeHandler: () => void;
+
+    private static device: Device;
+
+    private static getDevice(): Device {
+      if(!this.device) {
+        let device = new com.keyman.Device();
+        device.detect();
+
+        this.device = device;
+      }
+
+      return TouchAliasData.device;
```

Comment:
Why `TouchAliasData.device` and not `this.device`?

Reply:
Because I set the `device` field to `static`.  The issue is that I forgot to change the `if` statement's condition to reflect that - good catch.